src/util/README.md

```diff
@@ -101,19 +101,90 @@ of a while loop is a statement.
 
 \subsection symbolt_section symbolt, symbol_tablet, and namespacet
 
-To be documented.
-
-\subsubsection symbol_lifetimes_section Symbol lifetimes, symbol modes, name, base-name, pretty-name; semantic of lifetimes for symex?
-
-To be documented.
-
-\subsubsection storing_symbols_section Storing symbols and hiding symbols (namespacet)
-
-To be documented.
-
-\subsubsection ns_follow_section ns.follow
-
-To be documented.
+A symbol table is a mapping from symbol names to \ref symbolt objects, which
+store a symbol's name, attributes, type and perhaps value. They are used to
+describe local and global variables, type definitions and function prototypes
+and definitions.
+
+All symbols store a type (an instance of \typet). For function or method symbols
+these are \ref code_typet instances.
+
+Global variable symbols may have a value (an \ref exprt), in which case it is
+used to initialise the global. Method or function symbols may also have a value,
+in which case it is a \ref codet and gives the function definition. A method or
```

Comment:
It's the function definition until compiled away by `goto_convert`, at which point it will become `irept(ID_compiled)`.

Reply:
I've heard this said before, but I don't see any evidence of it actually being done?

Reply:
See #2717 - until then it's actually `irept("compiled")`.

Reply:
That appears to only happen in goto-cc; cbmc and so on will still carry its value. I'll add a note that this might happen, for certain driver programs.

Reply:
Done

Reply:
(It's actually an interesting question whether the code left in place when not using goto-cc consumes considerable memory - there will be very little sharing in that part.)